MetaMask/metamask-extension
# fix cached detected token results

~Blocked by https://github.com/MetaMask/controllers/pull/956 (and subsequent release of `@metamask/controllers`)~

Context (from https://github.com/MetaMask/controllers/pull/956):

_Currently there exists an issue (exposed by https://github.com/MetaMask/metamask-extension/issues/16349) where if a user changes accounts (account x to account y) or (changes from network a to network b) right as a detectTokens() call is occuring, the results of the getBalancesInSingleCall() call (withindetectTokens()) made against address x or network acan get incorrectly misattributed to accountyornetwork b. This can happen because currently detectTokens() simply calls this.addDetectedTokens() on the TokensController with the detected tokensToAdd assuming that when this call occurs in the TokensController the selectedAddress or chainId [used to store these detected assets](https://github.com/MetaMask/controllers/blob/9b82201b4c99059a7ab35942febdba840adce5ee/packages/assets-controllers/src/TokensController.ts#L709) will be the same as the one used to detect them. With sufficiently unfortunate timing this expectation can be broken, and so this PR introduces a pattern ([already used in the analogous flow between the NftDetectionController and the NftController](https://github.com/MetaMask/controllers/blob/main/packages/assets-controllers/src/NftDetectionController.ts#L397), whereby the address and network data are passed along with the detectedToken data to ensure they are stored correctly._

This PR leverages the changes made in https://github.com/MetaMask/controllers/pull/956, and additionally fixes a race condition that occurs where even if the detection data is stored correctly, the `detectedTokens` value does not get updated to reflect the correct chainId/network combo if/when the network/account is changed while the service worker is being restarted.

* Fixes https://github.com/MetaMask/metamask-extension/issues/16349


### Before



### After



## Manual Testing Steps

1. Unlock MM
2. Get some tokens on the first account, so they are detected
3. Create a fresh new account
4. Change back to the account from step 2
5. Terminate service worker
6. Change back to the account from step 3 -- see that no detected tokens are cached

## Pre-merge author checklist


## Pre-merge reviewer checklist


If further QA is required (e.g. new feature, complex testing steps, large refactor), add the `Extension QA Board` label.

In this case, a QA Engineer approval will be be required.
